File: client/nodejs/FastClient.js
/*
Info : FastClient in NodeJS for FastApi connection.
Author: Haseeb Mir.
*/

var request = require("request");
const { faker } = require("@faker-js/faker");

var url = "http://localhost:8000";

function ExecuteRequest(url, method, body) {
  var resp = null;
  var options = {
    method: method,
    url: url,
    headers: { "content-type": "application/json" },
    body: body,
    json: true,
  };

  request(options, function (error, response, data) {
    if (error) throw new Error(error);
    else console.log(data);
  });
  return resp;
}

//Show user.
function ShowUser(id) {
  var userUrl = url + "/user/?id=" + id;
  ExecuteRequest(userUrl, "GET", {});
}

//Show Users
function ShowUsers() {
  ExecuteRequest(url, "GET", {});
}

//Add user.
function AddUser(body) {
  var userUrl = url + "/add-user/";
  ExecuteRequest(userUrl, "POST", body);
}

//Update user.
function UpdateUser(id, name, age) {
  var userUrl =
    url + "/update-user/?id=" + id + "&name=" + name + "&age=" + age;
  ExecuteRequest(userUrl, "PATCH", {});
}

//Remove user.
function RemoveUser(id = 0) {
  var userUrl = url + "/remove-user/?id=" + id;
  ExecuteRequest(userUrl, "DELETE", {});
}

function CreateRandomUser() {
  const rand = faker.datatype.number();
  var gender = rand % 2 == 0 ? "male" : "female";
  var name = faker.name.firstName(gender);
  return {
    name: name,
    age: faker.datatype.number(100, 15, 100),
    gender:gender,
    email: name + "_" + faker.name.lastName(gender) + "@" + faker.internet.domainSuffix() + ".com",
    phone: "+" + faker.phone.number(),
    address: faker.address.country(),
  };
}

function GenerateRandUsers(length){
    var users = [];

    Array.from({ length: length }).forEach(() => {
      users.push(CreateRandomUser());
    });
    return users;
}

async function AddUsers(length) {
    var users = GenerateRandUsers(length);
    console.time('AddUsers');
    users.forEach((user) => {
      AddUser(user);
    });
    console.timeEnd('AddUsers');
}

//Main Function area..
function main() {
  
  //Adding user.
  var user = { name: "NodeJS-Pro", age: 22, gender: "M",email: "nodejs@pro.com",phone: "+971-54545400",address: "GBR"};
  AddUser(user)

  //Adding random users.
  //AddUsers(10)

  //Updating user.
  //UpdateUser(1,'GoPro',22)

  //Removing user.
  //RemoveUser(1)

  //Show single users.
  //ShowUser(1)

  //Show all users.
  ShowUsers()
}

if (require.main === module) {
  main();
}

File: client/golang/FastClient.go
/*
Info : FastClient in GoLang for FastApi connection.
Author: Haseeb Mir.
*/

package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"strconv"
	"strings"
)

var url = "http://localhost:8000"

func ExecuteRequest(url string, method string, payload string) string {
	var data = strings.NewReader(payload)
	var req, _ = http.NewRequest(method, url, data)
	var res, _ = http.DefaultClient.Do(req)

	defer res.Body.Close()
	var body, _ = ioutil.ReadAll(res.Body)
	return string(body)
}

func AddUser(name string, age int, gender string, email string, phone string, address string) string {
	var user_url = url + "/add-user/"
	var payload = "{\n\"name\":\"" + name + "\",\n\"age\":" + strconv.Itoa(age) + ",\n\"gender\":\"" + gender + "\",\n\"email\":\"" + email + "\",\n\"phone\":\"" + phone + "\",\n\"address\":\"" + address + "\"\n}"
	var resp = ExecuteRequest(user_url, "POST", payload)
	return resp
}

func GetUsers() string {
	var resp = ExecuteRequest(url, "GET", "")
	return resp
}

func UpdateUser(id int, name string, age int) string {
	var user_url = url + "/update-user/?id=" + strconv.Itoa(id) + "&name=" + name + "&age=" + strconv.Itoa(age)
	var resp = ExecuteRequest(user_url, "PATCH", "")
	return resp
}

func RemoveUser(id int) string {
	var user_url = url + "/remove-user/?id=" + strconv.Itoa(id)
	var resp = ExecuteRequest(user_url, "DELETE", "")
	return resp
}

func GetUser(id int) string {
	var user_url = url + "/user/?id=" + strconv.Itoa(id)
	var resp = ExecuteRequest(user_url, "GET", "")
	return resp
}

func main() {

	//Add user.
	var data = AddUser("GoLang", 13, "M", "goPro@gmail.com", "+971-545454", "GO")
	fmt.Println(data)

	//Update user.
	//var data = UpdateUser(0, "HaseebHM", 32)
	//fmt.Println(data)

	//Remove user.
	// var data = RemoveUser(1)
	// fmt.Println(data)

	//Show single user.
	//var data = GetUser(0)
	//fmt.Println(data)

	//Show all user.
	data = GetUsers()
	fmt.Println(data)
}
